README.md
# Cooldown

Timer mechanism to limit place upper bound on rate of events.

## Installation

    npm install cooldown

## Example

This example reads lines of text from stdin.  When 'spam' is entered, it will
output 'spam' but only at a rate of once per 5 seconds.  Entering 'reset' can
reset the cooldown and allow 'spam' to succeed immediately after.  Entering
'ready?' will display the state of the cooldown timer.  The 'ready' event
listener will automatically print when the timer is off cooldown.

```javascript
var Cooldown = require('cooldown');
var readline = require('readline');

var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Set limit to 5s
var cd = new Cooldown(5000);
cd.on('ready', console.log.bind('console', 'off cooldown'));

rl.on('line', function (line) {
  switch (line) {
  case 'spam':
    if (cd.fire()) {
      console.log('have some spam');
    } else {
      console.log('not yet');
    }
    break;
  case 'ready?':
    console.log(cd.ready ? 'yep' : 'nope');
    break;
  case 'reset':
    // reset the cooldown
    cd.reset();
    break;
  case 'quit':
    cd.destroy();
    rl.close();
    break;
  }
});
```

## API

### Class Cooldown

A cooldown timer with two states:

- ready: available to be fired
- on-cooldown: fired and waiting to become ready again

#### new Cooldown(timeout)

- timeout: cooldown timeout duration (in ms)

#### cooldown.fire()

Return true if the timer was ready (and puts it on-cooldown).

#### cooldown.reset(noEmit)

If timer is on-cooldown, reset it back to ready.

- noEmit: do not emit 'ready' event if timer was on-cooldown

#### cooldown.destroy()

Clear any timeouts and set timer to on-cooldown.
It will never enter the 'ready' state unless it is reset.

#### Property: ready

Contains true if the timer is off cooldown and available to fire, else false.

#### Event: 'ready'

Emitted whenever the timer comes back off cooldown.

#### Event: 'cooldown'

Emitted whenever the timer goes on cooldown.

## License

BSD


## Bugs

See <https://github.com/pfmooney/node-cooldown/issues>.


CHANGES.md
# Changes


## v0.1.0

Initial release


test/main.test.js
// Copyright 2014, Patrick Mooney.  All rights reserved.

var test = require('tape').test;
var util = require('util');


///--- GLOBALS
var Cooldown;
var TIMEOUT = 500;
var cd;

var timerStart;
function start() {
  timerStart = process.hrtime();
}
function elapsed() {
  var diff = process.hrtime(timerStart);
  return (diff[0] * 1000)  + (diff[1] / 1000000);
}

test('load', function (t) {
  Cooldown = require('../index');
  t.ok(Cooldown);
  t.end();
});

test('basic', function (t) {
  var interval;

  cd = new Cooldown(TIMEOUT);

  start();
  t.ok(cd.fire(), 'started');

  function done() {
    clearInterval(interval);
    cd.destroy();
    t.end();
  }

  interval = setInterval(function () {
    var diff = elapsed();
    if (cd.ready) {
      t.ok(diff > TIMEOUT, util.format('%d > %d', diff, TIMEOUT));
      done();
    } else if (diff > 2*TIMEOUT) {
      t.fail('too late');
      done();
    }
  }, 10);
});

test('fire during cd', function (t) {
  cd = new Cooldown(TIMEOUT);
  t.ok(cd.ready, 'ready');
  start();
  t.ok(cd.fire(), 'started');
  t.ok(!cd.ready, 'not ready while on cd');

  // attempt to re-fire cooldown after 1/2 timeout
  setTimeout(function () {
    t.ok(!cd.fire(), 'did not fire while on cd');
  }, TIMEOUT/2);

  cd.once('ready', function () {
    var diff = elapsed();
    var max = TIMEOUT + (TIMEOUT/2);
    t.ok(diff > TIMEOUT, util.format('%d > %d', diff, TIMEOUT));
    // ensure the second firing didn't stretch the window
    t.ok(diff < max, util.format('%d < %d', diff, max));
    t.end();
  });
});

test('events', function (t) {
  cd = new Cooldown(TIMEOUT);

  cd.once('cooldown', function () {
    t.ifError(cd.ready, 'not ready while on cd');
    cd.once('ready', function () {
      t.ok(cd.ready, 'ready when off cd');
      t.end();
    });
  });

  t.ok(cd.fire(), 'start');
});

test('reset', function (t) {
  cd = new Cooldown(TIMEOUT);

  start();
  t.ok(cd.fire(), 'started');
  cd.once('ready', function () {
    t.ok(cd.ready, 'reset');
    var diff = elapsed();
    t.ok(diff < TIMEOUT, util.format('%d < %d', diff, TIMEOUT));
    cd.on('ready', t.fail.bind(t));
    t.end();
  });
  cd.reset();
});

test('reset - already ready', function (t) {
  cd = new Cooldown(TIMEOUT);
  t.ok(cd.ready);
  cd.on('ready', t.fail.bind(t, 'ready should not be emitted'));
  cd.reset();
  t.ok(cd.ready);
  t.end();
});

test('reset - noEmit', function (t) {
  cd = new Cooldown(TIMEOUT);

  t.ok(cd.fire(), 'started');
  cd.on('ready', t.fail.bind(t, 'ready should not be emitted'));
  cd.reset(true);
  t.ok(cd.ready, 'reset');
  t.end();
});

test('destroy', function (t) {
  cd = new Cooldown(TIMEOUT);

  t.ok(cd.fire(), 'started');
  cd.on('ready', t.fail.bind(t, 'ready should not be emitted'));
  cd.destroy();
  t.ok(!cd.ready, 'destroyed');
  t.end();
});


index.js
// Copyright 2014, Patrick Mooney.  All rights reserved.

var util = require('util');
var EventEmitter = require('events').EventEmitter;

var assert = require('assert-plus');

/**
 * Manage a cooldown period.
 *
 * @param {Number} timeout - length in ms of cooldown periods.
 * @property {Boolean} ready - false if during a cooldown period.
 */
function Cooldown(timeout) {
  assert.number(timeout);
  assert.ok(timeout > 0);
  EventEmitter.call(this);

  var self = this;
  this._timeout = timeout;
  this._ready = true;

  this.__defineGetter__('ready', function () {
    return self._ready;
  });
}
util.inherits(Cooldown, EventEmitter);
module.exports = Cooldown;

/**
 * Attempt to fire the cooldown.
 *
 * @return {Boolean} false if during cooldown period, else true.
 */
Cooldown.prototype.fire = function fire() {
  if (this._ready) {
    this._ready = false;
    this._lastRun = process.hrtime();
    this._checkTimer(true);
    this.emit('cooldown');
    return true;
  }
  return false;
};

/**
 * Reset a cooldown period.
 *
 * @param {Number} noEmit do not emit ready if resetting cooldown.
 */
Cooldown.prototype.reset = function reset(noEmit) {
  if (!this._ready) {
    clearTimeout(this._timer);
    this._timer = null;
    this._ready = true;
    if (!noEmit) {
      this.emit('ready');
    }
  }
};

/**
 * Destroy cooldown object, clearing any timers.
 */
Cooldown.prototype.destroy = function destroy() {
  if (this._timer) {
    clearTimeout(this._timer);
  }
  this._ready = false;
};

/**
 * Set or re-set timer for cooldown window.
 *
 * To ensure that clock variations don't cause a cooldown period to end before
 * adequate time has elapsed, setTimeout is not trusted completely.  When it
 * ends, the elapsed interval is checked with process.hrtime.  If adequate time
 * has passed, it is indicated to consumers, otherwise another timeout to make
 * up the difference is initiated.
 */
Cooldown.prototype._checkTimer = function _checkTimer(initial) {
  // Use process.hrtime to ensure that the timeout elapses before indicating
  // the end of the cooldown.
  var elapsed = process.hrtime(this._lastRun);
  var msElapsed = (elapsed[0] * 1000) + (elapsed[1] / 1000000);
  if (msElapsed >= this._timeout && !initial) {
    this._ready = true;
    this._timer = null;
    this.emit('ready');
  } else {
    // It's not _quite_ time yet
    this._timer = setTimeout(this._checkTimer.bind(this),
        (this._timeout - msElapsed)|0);
  }
};
